Refactor FXIOS-9183 - Replace the Learn more button in Tracking Protection to use LinkButton in Component Library

bmihai23 opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-9183)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/20344)

## :bulb: Description
- This PR replaces the generic UIButton with the LinkButton from the ComponentLibrary in the ContentBlockerSettingViewController. 

**Before**
| Portrait  | Landscape | 
| ------------- | ------------- |
| ![Simulator Screenshot - iPhone 15 Pro - 2024-05-23 at 10 59 29](https://github.com/mozilla-mobile/firefox-ios/assets/74779930/0bdfd01d-5520-46f3-aa84-a556d50cddf0) | ![Simulator Screenshot - iPhone 15 Pro - 2024-05-23 at 16 38 25](https://github.com/mozilla-mobile/firefox-ios/assets/74779930/93ccd673-1868-4714-9b3e-99d13dddbef8) |

**After**
| Portrait  | Landscape | 
| ------------- | ------------- |
| ![Simulator Screenshot - iPhone 15 Pro - 2024-05-29 at 21 46 09](https://github.com/mozilla-mobile/firefox-ios/assets/74779930/10157fc4-5cc3-4d41-971b-cff225912f08) | ![Simulator Screenshot - iPhone 15 Pro - 2024-05-29 at 21 46 12](https://github.com/mozilla-mobile/firefox-ios/assets/74779930/bd1d3bc1-cda8-4bff-b1f5-3408ef9eff15) |
| ![Simulator Screenshot - iPhone 15 Pro - 2024-06-03 at 21 13 51](https://github.com/mozilla-mobile/firefox-ios/assets/74779930/ba848aa5-f50a-4605-97fc-a19c82ac0d2f) | ![Simulator Screenshot - iPhone 15 Pro - 2024-06-03 at 21 13 58](https://github.com/mozilla-mobile/firefox-ios/assets/74779930/ffcadbc1-ae69-429a-812a-88e777dea922) |


## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

cyndichin commented on firefox-ios/Client/Frontend/Settings/ContentBlockerSettingViewController.swift:
```diff
@@ -145,20 +151,29 @@ class ContentBlockerSettingViewController: SettingsTableViewController {
-            button.setAttributedTitle(NSAttributedString(string: title, attributes: attributes), for: .normal)
-            button.addTarget(self, action: #selector(moreInfoTapped), for: .touchUpInside)
-            button.isHidden = false
+            linkButton.setAttributedTitle(NSAttributedString(string: title, attributes: attributes), for: .normal)
```
thanks for this update!! Can we also use `LinkButtonViewModel` to configure the button? We want the new link buttons to have an underline style.

bmihai23 replied:
Hi @cyndichin, the button is now configured using `LinkButtonViewModel` and I also updated the screenshots. Thank you 🙏

cyndichin commented on firefox-ios/Client/Frontend/Settings/ContentBlockerSettingViewController.swift:
```diff
@@ -185,4 +202,10 @@ class ContentBlockerSettingViewController: SettingsTableViewController {
+    // MARK: - ThemeApplicable
+    func applyTheme(theme: Theme) {
+        let colors = theme.colors
+        linkButton.setTitleColor(colors.actionPrimary, for: .normal)
```
to keep consistent with other other theming, lets add `linkButton.applyTheme(theme: theme)` here instead of in its declaration and it should set the title automatically using the component.

cyndichin reviewed (commented):
thanks @bmihai23 for making the update using the view model! Looks great! Had one more clarifying comment and another comment for improvement.

bmihai23 commented on firefox-ios/Client/Frontend/Settings/ContentBlockerSettingViewController.swift:
```diff
@@ -145,20 +151,31 @@ class ContentBlockerSettingViewController: SettingsTableViewController {
-            button.isHidden = false
+            defaultFooter.contentView.addSubview(linkButton)
 
-            defaultFooter.stackView.addArrangedSubview(button)
```
I did this change because the text and the button wouldn’t align correctly even if I tried to add constraints. If you consider that I should keep it how it was initially, please let me know. Thanks 🙏🏻

cyndichin replied:
What about adjusting the `contentInsets` in the `LinkButtonViewModel` initializer, so that the leading and trailing is `0`? Let me know if that works!

cyndichin commented on firefox-ios/Client/Frontend/Settings/ContentBlockerSettingViewController.swift:
```diff
@@ -5,9 +5,15 @@
 
-class ContentBlockerSettingViewController: SettingsTableViewController {
-    private let button = UIButton()
+class ContentBlockerSettingViewController: SettingsTableViewController, ThemeApplicable {
```
since `SettingsTableViewController` is a `ThemedTableViewController`, we can remove `ThemeApplicable` and can override the `applyTheme` from the parent class.

bmihai23 commented:
Hi @cyndichin and thanks for your feedback, I just want to let you know that I updated the code using everything you said above and I also updated the screenshots. Thanks again 🙏

cyndichin commented on BrowserKit/Sources/ComponentLibrary/Buttons/LinkButtonViewModel.swift:
```diff
@@ -23,9 +23,9 @@ public struct LinkButtonViewModel {
                 font: UIFont = FXFontStyles.Regular.callout.scaledFont(),
                 contentInsets: NSDirectionalEdgeInsets = NSDirectionalEdgeInsets(top: UX.verticalInset,
-                                                                                 leading: UX.horizontalInset,
+                                                                                 leading: 0,
```
it looks like we're updating this for all our link buttons, but is only needed for this case. What do you think about passing this in when we initialize the `LinkButtonViewModel` for tracking protection?

cyndichin commented on firefox-ios/Client/Frontend/Settings/ContentBlockerSettingViewController.swift:
```diff
@@ -145,20 +149,19 @@ class ContentBlockerSettingViewController: SettingsTableViewController {
-            var attributes = [NSAttributedString.Key: AnyObject]()
-            attributes[NSAttributedString.Key.foregroundColor] = theme.colors.actionPrimary
-            attributes[NSAttributedString.Key.font] = font
+            linkButton.applyTheme(theme: currentTheme)
```
if we're applying the theme in `applyTheme` for this class, we can remove this here, which can remove creating the `currentTheme` at the class level. Instead we can create it in the local scope of the `applyTheme` method. Let me know if there's any issues with this.

cyndichin reviewed (commented):
thanks @bmihai23 for making the changes, looking a lot cleaner. I have a couple of more questions based on the latest changes. thank you for the work!

cyndichin commented on firefox-ios/Client/Frontend/Settings/ContentBlockerSettingViewController.swift:
```diff
@@ -185,4 +194,11 @@ class ContentBlockerSettingViewController: SettingsTableViewController {
+    // MARK: - ThemeApplicable
+    func applyTheme(theme: Theme) {
+        let colors = theme.colors
+        linkButton.setTitleColor(colors.actionPrimary, for: .normal)
```
we should be able to remove this now, and since we have an accessing `applyTheme`, what do you think about using the `override` method?

cyndichin commented on firefox-ios/Client/Frontend/Settings/ContentBlockerSettingViewController.swift:
```diff
@@ -185,4 +193,12 @@ class ContentBlockerSettingViewController: SettingsTableViewController {
+    override func applyTheme() {
+        super.applyTheme()
+        let currentTheme = currentTheme()
+        linkButton.setTitleColor(currentTheme.colors.actionPrimary, for: .normal)
```
thanks for the update! do we need this line here? If not, let's remove it so we can rely on the `applyTheme` for the link button.

bmihai23 replied:
Hi @cyndichin,  we don't need this line here because we don't want to have redundant pieces of code that are doing the same thing, title color is already set in the `LinkButton` class. Thank you 🙏

cyndichin reviewed (approved):
thanks @bmihai23 for working through the changes with me, it looks great now!

bmihai23 commented:
> thanks @bmihai23 for working through the changes with me, it looks great now!

It was my pleasure working with you, thank you for your continuous support 🙏🏻
